src/controllers/aer_controller.hpp

```diff
@@ -1404,85 +1525,262 @@ Transpile::Fusion Controller::transpile_fusion(Method method,
 template <class State_t>
 void Controller::run_circuit_helper(const Circuit &circ,
                                     const Noise::NoiseModel &noise,
-                                    const json_t &config, uint_t shots,
+                                    const json_t &config,uint_t shots,
                                     uint_t rng_seed, const Method method,
                                     bool cache_blocking,
-                                    ExperimentResult &result) const {
-  // Initialize new state object
-  State_t state;
+                                    ExperimentResult &result) const
+{
+  // Start individual circuit timer
+  auto timer_start = myclock_t::now(); // state circuit timer
 
-  // Validate gateset and memory requirements, raise exception if they're exceeded
-  validate_state(state, circ, noise, true);
+  // Initialize circuit json return
+  result.legacy_data.set_config(config);
 
-  // Set state config
-  state.set_config(config);
-  state.set_parallalization(parallel_state_update_);
-  state.set_distribution(num_processes_);
-  state.set_global_phase(circ.global_phase_angle);
+  // Execute in try block so we can catch errors and return the error message
+  // for individual circuit failures.
+  try {
+    // Rng engine (this one is used to add noise on circuit)
+    RngEngine rng;
+    rng.set_seed(rng_seed);
+
+    // Output data container
+    result.set_config(config);
+    result.metadata.add(method_names_.at(method), "method");
+    if (method == Method::statevector || method == Method::density_matrix ||
+        method == Method::unitary) {
+      result.metadata.add(sim_device_name_, "device");
+    } else {
+      result.metadata.add("CPU", "device");
+    }
 
-  // Rng engine
-  RngEngine rng;
-  rng.set_seed(rng_seed);
+    // Circuit qubit metadata
+    result.metadata.add(circ.num_qubits, "num_qubits");
+    result.metadata.add(circ.num_memory, "num_clbits");
+    result.metadata.add(circ.qubits(), "active_input_qubits");
+    result.metadata.add(circ.qubit_map(), "input_qubit_map");
+    result.metadata.add(circ.remapped_qubits, "remapped_qubits");
+
+    // Add measure sampling to metadata
+    // Note: this will set to `true` if sampling is enabled for the circuit
+    result.metadata.add(false, "measure_sampling");
+    result.metadata.add(false, "batched_shots_optimization");
+
+    if(circ.num_qubits > 0){  //do nothing for query steps
+      // Choose execution method based on noise and method
+      Circuit opt_circ;
```

Comment:
We can reduce duplication to generate a sampling circuit in `run_circuit_helper` and `run_batched_circuits_helper` by adding a new method:

```
void Controller::get_sampling_circuit(const Circuit &circ,
                                      const Noise::NoiseModel &noise,
                                      RngEngine &rng,
                                      const Method method,
                                      Circuit &sampling_circuit,
                                      std::string &sampling_method) const;
```

Reply:
created a function for noise sampling